jax_distribution/systems/ff_mappo_rware.py

```diff
@@ -0,0 +1,523 @@
+    # Add dimension to pmap over.
+    reshape_step_rngs = lambda x: x.reshape(
+        (n_devices, config["update_batch_size"]) + x.shape[1:]
+    )
+    step_rngs = reshape_step_rngs(jnp.stack(step_rngs))
+    reshape_states = lambda x: x.reshape(
+        (n_devices, config["update_batch_size"], config["num_envs"]) + x.shape[1:]
+    )
+    env_states = jax.tree_util.tree_map(reshape_states, env_states)
+    timesteps = jax.tree_util.tree_map(reshape_states, timesteps)
+
+    init_learner_state = LearnerState(
+        params, opt_state, step_rngs, env_states, timesteps
+    )
+    return learn, network, init_learner_state
+
+
+def run_experiment(_run: run.Run, _config: Dict, _log: SacredLogger) -> None:
+    """Runs experiment."""
+    # Logger setup
+    config = config_copy(_config)
+    log = logger_setup(_run, config, _log)
+
+    generator = RandomGenerator(**config["rware_scenario"])
+    # Create envs
+    env = jumanji.make(config["env_name"], generator=generator)
+    env = RwareMultiAgentWithGlobalStateWrapper(env)
+    env = AutoResetWrapper(env)
+    env = LogWrapper(env)
+    eval_env = jumanji.make(config["env_name"], generator=generator)
+    eval_env = RwareMultiAgentWithGlobalStateWrapper(eval_env)
+
+    # PRNG keys.
+    rng, rng_e, rng_p = jax.random.split(jax.random.PRNGKey(config["seed"]), num=3)
+    # Setup learner.
+    learn, network, learner_state = learner_setup(env, (rng, rng_p), config)
+
+    # Setup evaluator.
+    evaluator, absolute_metric_evaluator, (trained_params, eval_rngs) = evaluator_setup(
+        eval_env, rng_e, network, learner_state.params, config
+    )
+
+    # Calculate total timesteps.
+    n_devices = len(jax.devices())
+    config["num_updates_per_eval"] = config["num_updates"] // config["num_evaluation"]
+    timesteps_per_training = (
+        n_devices
+        * config["num_updates_per_eval"]
+        * config["rollout_length"]
+        * config["update_batch_size"]
+        * config["num_envs"]
+    )
+
+    # Run experiment for a total number of evaluations.
+    max_episode_return = jnp.float32(0.0)
+    best_params = None
+    for i in range(config["num_evaluation"]):
+        # Train.
+        with TimeIt(
+            tag=("COMPILATION" if i == 0 else "EXECUTION"),
+            environment_steps=timesteps_per_training,
+        ):
+            learner_output = learn(learner_state)
+            jax.block_until_ready(learner_output)
+
+        # Prepare for evaluation.
+        trained_params = jax.tree_util.tree_map(
+            lambda x: x[:, 0, ...], learner_output.learner_state.params
+        )
+        rng_e, *eval_rngs = jax.random.split(rng_e, n_devices + 1)
+        eval_rngs = jnp.stack(eval_rngs)
+        eval_rngs = eval_rngs.reshape(n_devices, -1)
+
+        # Evaluate.
+        evaluator_output = evaluator(trained_params, eval_rngs)
+        jax.block_until_ready(evaluator_output)
+
+        # Log the results
+        log(
```

Comment:
Minor detail, but we should log the learner output before the evaluation happens.

Reply:
Nice catch will update it for the ff-ippo too